rfcs/records.md

```diff
@@ -0,0 +1,196 @@
+
+```
+local X = require(path).X
+local r = new X(1, 2)
+```
+
+### Type checking records
+
+Records defined via a `record` statement can be used in type annotations as usual. The unification rules say that a table can unify with a record if the fields
+match, which makes records similar to sealed tables from the type checking perspective. (note, this is hand wavy on subtyping rules)
+
+> TODO: How do you export a record type? `export record` would be straightforward but potentially conflicts with future export statements for functions/values.
+
+The methods defined on the record object are type checked as usual, with one exception - the implicit `self` has the type of the record. This is crucial because
+this is the one big issue we can't resolve with metatable-based OOP for tables, but it works for records because `:` is slightly more magical.
+
+When `self` is explicit, the type needs to be specified manually, e.g. these definitions are equivalent:
+
+```
+function Point:sum(): number
+    return self.x + self.y
+end
+
+function Point.sum(self: Point): number
+    return self.x + self.y
+end
+```
+
+Of course, the type checker also knows that the record type has the metatable with the inferred type of the record shape.
+
+> TODO: Does the type checker need to understand the internal structure of the shape so that type checking works across modules, or is simply modeling this
+> as a metatable sufficient?
+
+### Object modeling
+
+> TODO: records don't support implementation inheritance and why it's a good thing
+
+## Drawbacks
+
+Adding a new data type that is cross-cutting (across syntax, semantics/compiler, semantics/type checking, and runtime) results in added complexity.
+
+The rigidity of records may make some applications hesitate to adopt them; e.g. you can't simply add a new field at a random point in the program, which some would
+argue makes the language less dynamic and therefore less convenient.
+
+Not enforcing type compatibility for typed records at runtime may make it difficult for us to optimize record storage more by removing the type tags (which could
+make record objects ~2x more efficient in some cases).
+
+## Alternatives
+
+Instead of using explicit record types, we can make the VM recognize shapes of objects automatically, just like JavaScript implementations do. This requires a
+substantial amount of complicated machinery and heuristics, and likely can't be as efficient as records in the long run, but it can result in close efficiency
+without any changes to existing programs. This, however, doesn't make type safety any easier.
+
+Instead of using record types that have minimal featureset, we could implement classes that have a more feature-rich OOP semantics, with inheritance, first class
+properties, and access control. This would better map to other high level languages like TypeScript/Python, but would make the language and runtime more complicated.
```

Comment:
I am wondering how likely it is that we add classes to Luau.  If we do add classes, then do records still make sense?  If they do still make sense, we would probably want to think about how records and classes relate to one another and if anything could/should be shared between them (e.g. similar syntax, or perhaps making classes a extension of records in some way).

Reply:
The way I'm thinking about this is "why would we ever add classes if we had records". I'm planning to write some thoughts along these lines in the "Object modeling" section. There's a few OOP features that records don't give you; some of them could be introduced later although I'd question the need, and some maybe we just live without and it's not a big deal.

Reply:
P.S. The way this RFC's design is structured is to disassociate the data type from methods a little bit while still providing a straightforward path to get the data and methods tightly bound; this seems to me like the optimal way to do OOP that actually is workable in large system.

Reply:
> I am wondering how likely it is that we add classes to Luau.

Personally, I would not like to see classes implemented as a first-class feature in Luau. And, honestly, I would not really like to see records either. I'd hate to see Luau moving further and further away from vanilla Lua with all these new datatypes and syntax, it just doesn't feel right. I like vectors and gradual typing, but records may be a bit too far and classes would be way too far. Records are somewhat acceptable, but classes immediately give me Python vibes, and I really don't like Python.

Reply:
The way I'm thinking about this is that we have a set of problems that we can't cleanly resolve with just tables, and records provide a minimal solution, in the spirit of Lua/Luau, to address them. I agree that to the extent possible we should keep records simple, which is part of why this uses the name records instead of classes, as classes may carry a lot of baggage and provide expectations for a large feature set that we aren't likely to get.

Reply:
Yeah we're on the same page. I have a section on this RFC that will address this once I actually fill it.

Reply:
In my opinion, the current way to create "classes" in vanilla Lua -- even ones that don't use inheritance -- is very messy and making it work for Luau typing is very annoying. Private members/fields (ones that are not shown within the exported type) are very annoying to implement and even just including them as public requires a lot of boilerplate.

I think the idea of implementing classes is a very nice addition and would make writing encapsulated code a lot easier. The way I see it is records are basically structs that can have member functions, and in a lot of ways this is similar to how classes work. If an opt-in idiom (such as by using some setmetatable trick) is added such that inheritance can be used if the user so wishes, but not directly supported, that would be a greatly appreciated feature. That being said, that is not entirely necessary.

My only concern is that "record" semantically may not make a lot of sense for instantiable objects that work on code such as lexers or parser or game logic instances. (Sorry if this made no sense.) Basically, I wouldn't call a lexer class a "record" or a struct, but that's just semantics and -- in implementation -- records would probably fill in the job just fine.

IMO Luau records == Structs/basic form of classes.